rig-postgres/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+use std::fmt::Display;
+
+use rig::{
+    embeddings::{Embedding, EmbeddingModel},
+    vector_store::{VectorStoreError, VectorStoreIndex},
+    Embed, OneOrMany,
+};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use serde_json::Value;
+use sqlx::PgPool;
+use uuid::Uuid;
+
+pub struct PostgresVectorStore<Model: EmbeddingModel> {
+    model: Model,
+    pg_pool: PgPool,
+    documents_table: String,
+    distance_function: PgVectorDistanceFunction,
+}
+
+/* PgVector supported distances
+<-> - L2 distance
+<#> - (negative) inner product
+<=> - cosine distance
+<+> - L1 distance (added in 0.7.0)
+<~> - Hamming distance (binary vectors, added in 0.7.0)
+<%> - Jaccard distance (binary vectors, added in 0.7.0)
+ */
+pub enum PgVectorDistanceFunction {
+    L2,
+    InnerProduct,
+    Cosine,
+    L1,
+    Hamming,
+    Jaccard,
+}
+
+impl Display for PgVectorDistanceFunction {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            PgVectorDistanceFunction::L2 => write!(f, "<->"),
+            PgVectorDistanceFunction::InnerProduct => write!(f, "<#>"),
+            PgVectorDistanceFunction::Cosine => write!(f, "<=>"),
+            PgVectorDistanceFunction::L1 => write!(f, "<+>"),
+            PgVectorDistanceFunction::Hamming => write!(f, "<~>"),
+            PgVectorDistanceFunction::Jaccard => write!(f, "<%>"),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, sqlx::FromRow)]
+pub struct SearchResult {
+    id: Uuid,
+    document: Value,
+    //embedded_text: String,
+    distance: f64,
+}
+
+impl SearchResult {
+    pub fn into_result<T: DeserializeOwned>(self) -> Result<(f64, String, T), VectorStoreError> {
+        let document: T =
+            serde_json::from_value(self.document).map_err(VectorStoreError::JsonError)?;
+        Ok((self.distance, self.id.to_string(), document))
+    }
+}
+
+impl<Model: EmbeddingModel> PostgresVectorStore<Model> {
+    pub fn new(
+        model: Model,
+        pg_pool: PgPool,
+        documents_table: Option<String>,
+        distance_function: PgVectorDistanceFunction,
+    ) -> Self {
+        Self {
+            model,
+            pg_pool,
+            documents_table: documents_table.unwrap_or(String::from("documents")),
+            distance_function,
+        }
+    }
+
+    pub fn default(model: Model, pg_pool: PgPool) -> Self {
+        Self::new(model, pg_pool, None, PgVectorDistanceFunction::Cosine)
+    }
```

Comment:
Perhaps rename to `with_defaults`? `default` is a method name usually reserved for the `Default` trait implementation

Reply:
true, done